File: apps/android/app/src/main/java/ai/openclaw/app/chat/ChatSwarmProgress.kt
package ai.openclaw.app.chat

import kotlinx.serialization.json.JsonElement
import kotlinx.serialization.json.JsonObject
import kotlinx.serialization.json.JsonPrimitive

private const val MAX_TRACKED_SWARM_GROUPS = 10_000
private const val MAX_TRACKED_SWARM_CHILDREN = 100_000
private const val MAX_SWARM_PAGE_REQUESTS = 100
private const val MAX_RENDERED_SWARM_DOTS_PER_PHASE = 256

enum class ChatSwarmDotStatus {
  Running,
  Queued,
  Failed,
  Done,
  ;

  val label: String
    get() = name
}

data class ChatSwarmDot(
  val key: String,
  val label: String,
  val status: ChatSwarmDotStatus,
)

data class ChatSwarmPhase(
  val key: String,
  val title: String?,
  val dots: List<ChatSwarmDot>,
  val hidden: Int,
)

data class ChatSwarmGroup(
  val groupId: String,
  val label: String,
  val running: Int,
  val done: Int,
  val failed: Int,
  val narrator: String?,
  val phases: List<ChatSwarmPhase>,
)

internal data class ChatSwarmSessionPage(
  val sessions: List<ChatSessionEntry>,
  val totalCount: Int?,
  val nextOffset: Int?,
  val hasMore: Boolean?,
)

internal suspend fun collectChatSwarmChildSessions(
  fetchPage: suspend (Int) -> ChatSwarmSessionPage,
): List<ChatSessionEntry> {
  val rowsByKey = linkedMapOf<String, ChatSessionEntry>()
  var remainingPageRequests = MAX_SWARM_PAGE_REQUESTS
  repeat(4) {
    val rowsBeforePass = rowsByKey.size
    var expectedTotal: Int? = null
    val seenOffsets = mutableSetOf<Int>()
    var offset = 0
    while (
      remainingPageRequests > 0 &&
      rowsByKey.size < MAX_TRACKED_SWARM_CHILDREN &&
      seenOffsets.add(offset)
    ) {
      remainingPageRequests -= 1
      val page = fetchPage(offset)
      expectedTotal = page.totalCount
      for (row in page.sessions) {
        rowsByKey[row.key] = row
        if (rowsByKey.size >= MAX_TRACKED_SWARM_CHILDREN) break
      }
      if (rowsByKey.size >= MAX_TRACKED_SWARM_CHILDREN) return rowsByKey.values.toList()
      val hasMore = page.hasMore ?: expectedTotal?.let { offset + page.sessions.size < it } ?: false
      val nextOffset = page.nextOffset ?: (offset + page.sessions.size)
      if (!hasMore || page.sessions.isEmpty() || nextOffset <= offset) break
      offset = nextOffset
    }
    val added = rowsByKey.size - rowsBeforePass
    if (
      remainingPageRequests == 0 ||
      added == 0 ||
      expectedTotal?.let { rowsByKey.size >= it } != false
    ) {
      return rowsByKey.values.toList()
    }
  }
  return rowsByKey.values.toList()
}

internal fun chatSwarmEventBelongsToParent(
  payload: JsonObject,
  matchesParent: (String) -> Boolean,
): Boolean {
  val source = payload["session"].asObjectOrNull() ?: payload
  val explicitParent =
    normalizedSwarmValue(source["parentSessionKey"].asStringOrNull())
      ?: normalizedSwarmValue(source["spawnedBy"].asStringOrNull())
  if (explicitParent != null) return matchesParent(explicitParent)

  val kind = normalizedSwarmValue(payload["kind"].asStringOrNull())
  if (kind == "phase" || kind == "log") {
    val sessionKey = normalizedSwarmValue(payload["sessionKey"].asStringOrNull()) ?: return false
    return matchesParent(sessionKey)
  }

  val groupId =
    normalizedSwarmValue(payload["swarmGroupId"].asStringOrNull())
      ?: normalizedSwarmValue(source["swarmGroupId"].asStringOrNull())
      ?: return false
  if (!groupId.startsWith("swarm:")) return false
  val generatedParent = groupId.removePrefix("swarm:").substringBeforeLast(':', missingDelimiterValue = "")
  return generatedParent.isNotEmpty() && matchesParent(generatedParent)
}

private fun normalizedSwarmValue(value: String?): String? = value?.trim()?.takeIf(String::isNotEmpty)

internal class ChatSwarmActivityTracker {
  private val currentPhaseByGroup = linkedMapOf<String, String>()
  private val phaseRankByGroupPhase = linkedMapOf<String, Int>()
  private var nextPhaseRank = 0
  private val latestLogByGroup = linkedMapOf<String, String>()
  private val phaseByChild = linkedMapOf<String, String>()

  fun clear() {
    currentPhaseByGroup.clear()
    phaseRankByGroupPhase.clear()
    nextPhaseRank = 0
    latestLogByGroup.clear()
    phaseByChild.clear()
  }

  fun observe(payload: JsonObject): Boolean {
    val source = payload["session"].asObjectOrNull() ?: payload
    val groupId = normalized(payload["swarmGroupId"].asStringOrNull()) ?: normalized(source["swarmGroupId"].asStringOrNull()) ?: return false
    val kind = normalized((if ("kind" in payload) payload["kind"] else source["kind"]).asStringOrNull())
    val text = normalized((if ("text" in payload) payload["text"] else source["text"]).asStringOrNull())
    if ((kind == "phase" || kind == "log") && text != null) {
      if (kind == "phase") {
        val rankKey = phaseRankKey(groupId, text)
        if (rankKey !in phaseRankByGroupPhase) {
          setBounded(phaseRankByGroupPhase, rankKey, nextPhaseRank++, MAX_TRACKED_SWARM_CHILDREN)
        }
        setBounded(currentPhaseByGroup, groupId, text, MAX_TRACKED_SWARM_GROUPS)
      } else {
        setBounded(latestLogByGroup, groupId, text, MAX_TRACKED_SWARM_GROUPS)
      }
      return true
    }

    val childKey = normalized(source["key"].asStringOrNull()) ?: normalized(payload["sessionKey"].asStringOrNull()) ?: return true
    val explicitPhase = normalized(source["swarmPhase"].asStringOrNull()) ?: normalized(payload["swarmPhase"].asStringOrNull())
    if (explicitPhase != null) {
      setBounded(phaseByChild, childKey, explicitPhase, MAX_TRACKED_SWARM_CHILDREN)
      return true
    }
    if (payload["reason"].asStringOrNull() == "create" && childKey !in phaseByChild) {
      currentPhaseByGroup[groupId]?.let { phase ->
        setBounded(phaseByChild, childKey, phase, MAX_TRACKED_SWARM_CHILDREN)
      }
    }
    return true
  }

  fun decorate(rows: List<ChatSessionEntry>): List<ChatSessionEntry> =
    rows.map { row ->
      val groupId = normalized(row.swarmGroupId) ?: return@map row
      val phase = phaseByChild[row.key] ?: row.swarmPhase
      row.copy(
        swarmPhase = phase,
        swarmPhaseRank = phase?.let { phaseRankByGroupPhase[phaseRankKey(groupId, it)] } ?: row.swarmPhaseRank,
        swarmLog = latestLogByGroup[groupId] ?: row.swarmLog,
      )
    }

  private fun phaseRankKey(
    groupId: String,
    phase: String,
  ): String = "${groupId.length}:$groupId$phase"

  private fun normalized(value: String?): String? = value?.trim()?.takeIf(String::isNotEmpty)

  private fun <V> setBounded(
    values: LinkedHashMap<String, V>,
    key: String,
    value: V,
    limit: Int,
  ) {
    if (key !in values && values.size >= limit) {
      values.entries
        .firstOrNull()
        ?.key
        ?.let(values::remove)
    }
    values[key] = value
  }
}

internal fun buildChatSwarmGroups(
  sessions: List<ChatSessionEntry>,
  matchesParent: (String) -> Boolean,
): List<ChatSwarmGroup> {
  data class Entry(
    val phase: String?,
    val phaseRank: Int,
    val log: String?,
    val dot: ChatSwarmDot,
  )

  val byGroup = linkedMapOf<String, MutableList<Entry>>()
  sessions.forEach { row ->
    val groupId = row.swarmGroupId?.trim()?.takeIf(String::isNotEmpty) ?: return@forEach
    if (!belongsToParent(row, groupId, matchesParent)) return@forEach
    val status = swarmDotStatus(row) ?: return@forEach
    val label = listOfNotNull(row.label, row.displayName, row.derivedTitle, row.key).firstOrNull { it.isNotBlank() } ?: row.key
    byGroup.getOrPut(groupId, ::mutableListOf) +=
      Entry(
        phase = row.swarmPhase?.trim()?.takeIf(String::isNotEmpty),
        phaseRank = row.swarmPhaseRank ?: Int.MAX_VALUE,
        log = row.swarmLog?.trim()?.takeIf(String::isNotEmpty),
        dot = ChatSwarmDot(key = row.key, label = label, status = status),
      )
  }

  return byGroup
    .map { (groupId, entries) ->
      val phases = linkedMapOf<String, Triple<String?, Int, MutableList<ChatSwarmDot>>>()
      entries.forEach { entry ->
        val phaseKey = entry.phase ?: ""
        val existing = phases[phaseKey]
        if (existing == null) {
          phases[phaseKey] = Triple(entry.phase, entry.phaseRank, mutableListOf(entry.dot))
        } else {
          existing.third += entry.dot
          phases[phaseKey] = Triple(existing.first, minOf(existing.second, entry.phaseRank), existing.third)
        }
      }
      val projectedPhases =
        phases
          .map { (key, bucket) ->
            val dots =
              if (bucket.third.size > MAX_RENDERED_SWARM_DOTS_PER_PHASE) {
                bucket.third.sortedBy(ChatSwarmDot::status)
              } else {
                bucket.third
              }
            ChatSwarmPhase(
              key = key,
              title = bucket.first,
              dots = dots.take(MAX_RENDERED_SWARM_DOTS_PER_PHASE),
              hidden = (dots.size - MAX_RENDERED_SWARM_DOTS_PER_PHASE).coerceAtLeast(0),
            )
          }.sortedWith(compareBy<ChatSwarmPhase> { phases[it.key]?.second ?: Int.MAX_VALUE }.thenBy(ChatSwarmPhase::key))
      val dots = entries.map(Entry::dot)
      ChatSwarmGroup(
        groupId = groupId,
        label = groupId.substringAfterLast(':'),
        running = dots.count { it.status == ChatSwarmDotStatus.Running },
        done = dots.count { it.status == ChatSwarmDotStatus.Done },
        failed = dots.count { it.status == ChatSwarmDotStatus.Failed },
        narrator = entries.firstNotNullOfOrNull(Entry::log),
        phases = projectedPhases,
      )
    }.filter { group ->
      group.phases.any { phase -> phase.dots.any { it.status == ChatSwarmDotStatus.Queued || it.status == ChatSwarmDotStatus.Running } }
    }.sortedBy(ChatSwarmGroup::groupId)
}

private fun swarmDotStatus(row: ChatSessionEntry): ChatSwarmDotStatus? =
  when {
    row.status == "running" || row.hasActiveRun == true -> ChatSwarmDotStatus.Running
    row.status == "done" -> ChatSwarmDotStatus.Done
    row.status == "failed" || row.status == "killed" || row.status == "timeout" -> ChatSwarmDotStatus.Failed
    row.subagentRunState == "active" || row.hasActiveSubagentRun == true -> ChatSwarmDotStatus.Queued
    else -> null
  }

private fun belongsToParent(
  row: ChatSessionEntry,
  groupId: String,
  matchesParent: (String) -> Boolean,
): Boolean {
  if (row.parentSessionKey?.let(matchesParent) == true || row.spawnedBy?.let(matchesParent) == true) return true
  val parts = groupId.split(':')
  return parts.size > 2 && matchesParent(parts.drop(1).dropLast(1).joinToString(":"))
}

private fun JsonElement?.asObjectOrNull(): JsonObject? = this as? JsonObject

private fun JsonElement?.asStringOrNull(): String? =
  (this as? JsonPrimitive)
    ?.takeIf { it.isString }
    ?.content

File: ui/src/pages/chat/components/chat-swarm-progress.ts
import { html, nothing, type TemplateResult } from "lit";
import type { GatewaySessionRow } from "../../../api/types.ts";
import { t } from "../../../i18n/index.ts";
import { areUiSessionKeysEquivalent } from "../../../lib/sessions/session-key.ts";

type SwarmDotStatus = "queued" | "running" | "done" | "failed";

const SWARM_DOT_STATUS_RANK = { running: 0, queued: 1, failed: 2, done: 3 } as const;
const MAX_RENDERED_DOTS_PER_PHASE = 256;

type SwarmDot = {
  key: string;
  label: string;
  status: SwarmDotStatus;
};

type SwarmPhase = {
  title?: string;
  dots: SwarmDot[];
  hidden: number;
};

type SwarmGroup = {
  groupId: string;
  label: string;
  running: number;
  done: number;
  failed: number;
  narrator?: string;
  phases: SwarmPhase[];
};

type SwarmPhaseCarrier = {
  swarmLog?: unknown;
  swarmPhase?: unknown;
  swarmPhaseRank?: unknown;
};

function swarmStatusLabel(status: SwarmDotStatus): string {
  if (status === "queued") {
    return t("tasksPage.status.queued");
  }
  if (status === "running") {
    return t("tasksPage.status.running");
  }
  if (status === "done") {
    return t("activity.status.done");
  }
  return t("tasksPage.status.failed");
}

function swarmDotStatus(row: GatewaySessionRow): SwarmDotStatus | null {
  if (row.status === "running" || row.hasActiveRun === true) {
    return "running";
  }
  if (row.status === "done") {
    return "done";
  }
  if (row.status === "failed" || row.status === "killed" || row.status === "timeout") {
    return "failed";
  }
  return row.subagentRunState === "active" ? "queued" : null;
}

function groupTail(groupId: string): string {
  return groupId.split(":").findLast(Boolean) ?? groupId;
}

function swarmPhaseRank(row: GatewaySessionRow): number {
  const rank = (row as GatewaySessionRow & SwarmPhaseCarrier).swarmPhaseRank;
  return typeof rank === "number" && Number.isFinite(rank) ? rank : Number.MAX_SAFE_INTEGER;
}

function swarmPhase(row: GatewaySessionRow): string | undefined {
  const phase = (row as GatewaySessionRow & SwarmPhaseCarrier).swarmPhase;
  return typeof phase === "string" && phase.trim() ? phase.trim() : undefined;
}

function swarmLog(row: GatewaySessionRow): string | undefined {
  const log = (row as GatewaySessionRow & SwarmPhaseCarrier).swarmLog;
  return typeof log === "string" && log.trim() ? log.trim() : undefined;
}

function isSwarmChildForSession(row: GatewaySessionRow, sessionKey: string): boolean {
  if (
    (row.parentSessionKey && areUiSessionKeysEquivalent(row.parentSessionKey, sessionKey)) ||
    (row.spawnedBy && areUiSessionKeysEquivalent(row.spawnedBy, sessionKey))
  ) {
    return true;
  }
  const owner = row.swarmGroupId?.split(":").slice(1, -1).join(":");
  return Boolean(owner && areUiSessionKeysEquivalent(owner, sessionKey));
}

function collectActiveSwarmGroups(
  sessions: readonly GatewaySessionRow[],
  sessionKey: string,
): SwarmGroup[] {
  const byGroup = new Map<
    string,
    Array<{ phase?: string; phaseRank: number; log?: string; dot: SwarmDot }>
  >();
  for (const row of sessions) {
    const groupId = row.swarmGroupId?.trim();
    if (!groupId || !isSwarmChildForSession(row, sessionKey)) {
      continue;
    }
    const status = swarmDotStatus(row);
    if (!status) {
      continue;
    }
    const entries = byGroup.get(groupId) ?? [];
    entries.push({
      phase: swarmPhase(row),
      phaseRank: swarmPhaseRank(row),
      log: swarmLog(row),
      dot: {
        key: row.key,
        label: row.label?.trim() || row.displayName?.trim() || row.derivedTitle?.trim() || row.key,
        status,
      },
    });
    byGroup.set(groupId, entries);
  }

  return [...byGroup.entries()]
    .map(([groupId, entries]) => {
      const dots = entries.map((entry) => entry.dot);
      const phases = new Map<string | undefined, { rank: number; dots: SwarmDot[] }>();
      for (const entry of entries) {
        const bucket = phases.get(entry.phase) ?? { rank: entry.phaseRank, dots: [] };
        bucket.rank = Math.min(bucket.rank, entry.phaseRank);
        bucket.dots.push(entry.dot);
        phases.set(entry.phase, bucket);
      }
      return {
        groupId,
        label: groupTail(groupId),
        running: dots.filter((dot) => dot.status === "running").length,
        done: dots.filter((dot) => dot.status === "done").length,
        failed: dots.filter((dot) => dot.status === "failed").length,
        narrator: entries.map((entry) => entry.log).find(Boolean),
        phases: [...phases.entries()]
          .toSorted((left, right) => left[1].rank - right[1].rank)
          .map(([title, bucket]) => {
            const visibleFirst =
              bucket.dots.length > MAX_RENDERED_DOTS_PER_PHASE
                ? bucket.dots.toSorted(
                    (left, right) =>
                      SWARM_DOT_STATUS_RANK[left.status] - SWARM_DOT_STATUS_RANK[right.status],
                  )
                : bucket.dots;
            return {
              title,
              dots: visibleFirst.slice(0, MAX_RENDERED_DOTS_PER_PHASE),
              hidden: Math.max(0, visibleFirst.length - MAX_RENDERED_DOTS_PER_PHASE),
            };
          }),
      } satisfies SwarmGroup;
    })
    .filter((group) =>
      group.phases.some((phase) =>
        phase.dots.some((dot) => dot.status === "queued" || dot.status === "running"),
      ),
    )
    .toSorted((left, right) => left.groupId.localeCompare(right.groupId));
}

export function renderChatSwarmProgress({
  sessions,
  sessionKey,
}: {
  sessions: readonly GatewaySessionRow[];
  sessionKey: string;
}): TemplateResult | typeof nothing {
  const groups = collectActiveSwarmGroups(sessions, sessionKey);
  if (groups.length === 0) {
    return nothing;
  }
  return html`
    <aside
      class="chat-swarm"
      data-test-id="chat-swarm"
      role="status"
      aria-live="off"
      aria-label=${t("labsPage.swarm.title")}
    >
      ${groups.map(
        (group) => html`
          <div class="chat-swarm__group" data-swarm-group=${group.groupId}>
            <div class="chat-swarm__header">
              <strong title=${group.groupId}>${group.label}</strong>
              <span
                >${group.running} ${swarmStatusLabel("running")} · ${group.done}
                ${swarmStatusLabel("done")} · ${group.failed} ${swarmStatusLabel("failed")}</span
              >
            </div>
            ${group.narrator
              ? html`<div class="chat-swarm__narrator">${group.narrator}</div>`
              : nothing}
            ${group.phases.map(
              (phase) => html`
                <div class="chat-swarm__phase-row">
                  <div class="chat-swarm__phase">
                    ${phase.title ?? t("labsPage.swarm.defaultPhase")}
                  </div>
                  <div class="chat-swarm__dots" role="list">
                    ${phase.dots.map(
                      (dot) => html`
                        <span
                          class=${`chat-swarm__dot chat-swarm__dot--${dot.status}`}
                          role="listitem"
                          title=${`${dot.label}: ${swarmStatusLabel(dot.status)}`}
                          aria-label=${`${dot.label}: ${swarmStatusLabel(dot.status)}`}
                        ></span>
                      `,
                    )}
                    ${phase.hidden > 0
                      ? html`<span class="chat-swarm__more" role="listitem">+${phase.hidden}</span>`
                      : nothing}
                  </div>
                </div>
              `,
            )}
          </div>
        `,
      )}
    </aside>
  `;
}
